### ISSUE TITLE
No data in Shelly components

### ISSUE THREAD

**Issue by gergwo** [Vanuan/switchbot-meter-to-shelly#2]
After following all detailed steps I am not getting any data from Switchbot meter on Shelly components.

**Comment by Vanuan** [comment]
Could you show the output of this screen?
![image](https://github.com/user-attachments/assets/10ff092a-46e5-4cc9-a3ae-ed95d76ed5e7)

**Comment by gergwo** [comment]
Hi Vanuan, 
I have found that the Switchbot MAC address has to be in capital letters, now in Shelly console I see temp, and humidity but there is still and error on the script. I will send you a picture.
![7A075915-6731-4B91-A893-EAF660AB30F6](https://github.com/user-attachments/assets/6444f308-da7d-49e3-8d2e-2fa3ce1f041d)

**Comment by gergwo** [comment]
Sorry this is the previous picture before correcting MAC address in capital letters.
![Uploading 0B914341-60E7-4C02-A256-F0016485FEB4.png…]()

**Comment by gergwo** [comment]
![B4C673AA-7184-4BEA-85A8-4B1026B2F05C](https://github.com/user-attachments/assets/7bce87d8-939a-419f-b16d-c2221420f4cd)

**Comment by gergwo** [comment]
![AEB6899B-D057-43FB-AB75-82455801B0F2](https://github.com/user-attachments/assets/66fae993-0219-4b34-a4d6-eac2f4597e9f)

**Comment by gergwo** [comment]
As you can see the error says that virtual is not defined. How can I fix it? I still do not see temperature and humidity in components. I think it is due this error. Could you please help me out to solve this issue? Thanks a lot.

**Comment by gergwo** [comment]
![Uploading B0C2FAD6-45A4-4088-8832-BDA28F2FF89D.png…]()

**Comment by gergwo** [comment]
Vanuan I have a question. Do you think that it could be difficult to adapt your code to read values from Switchbot pro meter which has CO2, temp, and humidity?I want to follow your example to try to get those values from that sensor. I have little experience in JS, but I want to make a try.

**Comment by gergwo** [comment]
This is the whole console error I get. I have been trying to fix it without success. 
![379C5154-5DB4-4359-BB1D-F99D12DD0495](https://github.com/user-attachments/assets/9588b739-94c9-4047-889b-b654ce676800)

**Comment by Vanuan** [comment]
Could you check this screen:


![Screenshot_20241221-183949](https://github.com/user-attachments/assets/d01363b6-4fc3-45b2-a1d3-35800638b932)

What do you see there? 

What Shelly device do you have? If you have 1st generation device, and firmware 0.14, those don't support Virtual components, unfortunately. Only some Gen2 and Gen3 Shelly relays have Virtual components. They first appeared in firmware 1.1.0-beta3

**Comment by gergwo** [comment]
My devices are gen3, updated to last firmware . What is the reason of that error. In components I do not see any value except zero for temp and humidity

**Comment by gergwo** [comment]
![B828BD23-6599-436A-9659-06C1C74E2DC5](https://github.com/user-attachments/assets/e9fd68e2-a591-4e23-8d46-5310deb49d80)

**Comment by gergwo** [comment]
This I from my Shelly app, but in web ui components are not showing values shown in console

**Comment by gergwo** [comment]
Hi Vanuan. You script is awesome. Thank you for sharing all your knowledge to all community. I finally succeeded getting the meter data in Shelly components. I updated it a little bit to get temperature in Fahrenheit degrees. 
Now I dealing with Switchbot pro meter to try to extract CO2 data, but as I mentioned before I am not an expert in JS. Have you tried that sensor? How can I find the instruction to get that data from such a sensor? I saw in your program that you extract temperature, and humidity from certain specific bits, or bytes.
![3D994C55-D4C9-4DFD-9D8D-7987E7D78ACF](https://github.com/user-attachments/assets/3c8134ba-2324-4f5a-81fb-36dca1cb1b9a)

**Comment by gergwo** [comment]
I have developed some sensors (like differential pressure sensors, temperature and humidity sensors) using Sensirion products. That company manufactures the best products you can find anywhere. That is the reason I am interested in Switchbot since they use sensirion inside their gadgets. Nowadays I am using a NDIR CO2 sensor read by  Shelly app, but I would really like to get data from the ones from Sensirion. After COVID I have realized that a CO2 sensor is handy to monitor environments around us  to prevent being infected by viruses, etc. 
As another person suggested you, you should publish your script in any Shelly contest. I bet your script would be welcome, and fully recognized by Shelly fans, and company CEO Dimitar Dimitrov, who from time to time I am in contact to.
With your modified  script to read CO2 anybody could benefit to make an automatic system to replete outdoor air in, filter it, and condition it through an AC unit, thus renovating the indoor air that is 5 or more times polluted than outdoor air.

**Comment by gergwo** [comment]
I have developed some sensors (like differential pressure sensors, temperature and humidity sensors) using Sensirion products. That company manufactures the best products you can find anywhere. That is the reason I am interested in Switchbot since they use sensirion inside their gadgets. Nowadays I am using a NDIR CO2 sensor read by Shelly app, but I would really like to get data from the ones from Sensirion. After COVID I have realized that a CO2 sensor is handy to monitor environments around us to prevent being infected by viruses, etc.
As another person suggested you, you should publish your script in any Shelly contest. I bet your script would be welcome, and fully recognized by Shelly fans, and company CEO Dimitar Dimitrov, who from time to time I am in contact to.
With your modified script to read CO2 anybody could benefit to make an automatic system to let outdoor air in, filter it, and condition it through an AC unit, thus renovating the indoor air that is 5 or more times polluted than outdoor air.

**Comment by Vanuan** [comment]
It seems like you solved the original issue. Could you care to explain what was the culprit? Might also worth sharing what are the changes to the script to support Fahrenheit. Was it only labeling or you did some math for conversion?

I've created another issue for the CO2, let's continue there on that: https://github.com/Vanuan/switchbot-meter-to-shelly/issues/3

**Comment by gergwo** [comment]
This script works perfectly after last update by creator. Thanks.

**Comment by gergwo** [comment]
Hi John, 
For the Fahrenheit convention I just added the math formula to the variable. 
You fixed the issue when you updated the script for the other guy who had the same problem than myself. I just copied your updated script, it started working in Shelly components. 
What  alsoI changed in my code it was the MAC address that it was not in capital letters at first. Then I updated script, and everything was fine. Thanks again. 
Remember if you want me to send you a Switchbot pro meter with CO2 data let me know, and give me the direction to send it. 
Happy Merry Christmas

Bests Regards,

Gerardo Ortiz
